### PR TITLE
py-cython: support Python 3.12+

### PR DESCRIPTION
**Pull request by adamjstewart**
Reboot of #46226, borrowing some ideas from #46547

Closes #46547

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-cython/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cython/package.py:21]: This URL isn't actually correct, it's missing a `/c/` in the path. Not sure how this ever worked.

File: var/spack/repos/builtin/packages/py-cython/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/py-cython/package.py]: I would drop all redundant lowerbounds, they just make it hard to read that the goal is to specify [forward compat bounds](https://spack.readthedocs.io/en/latest/packaging_guide.html#specifying-backward-and-forward-compatibility).
```suggestion
        depends_on("python@:3.13")  # or maybe even drop this line until 3.14 is released
        depends_on("python@:3.12", when="@:3.0.10")
        depends_on("python@:3.11", when="@:0.29.34")
        depends_on("python@:3.10", when="@:0.29.28")
        depends_on("python@:3.9", when="@:0.29.24")
        depends_on("python@:3.8", when="@:0.29.20")
```

File: var/spack/repos/builtin/packages/py-cython/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/py-cython/package.py]: why don't you make this another forward compat bound in the same style?

```
depends_on("python@:3.11", when="@:3.0.3")
```

or update the when condition from `@:0.29.34` to `@:3.0.3`

**Comment by haampie** [comment]
thanks for doing this :+1:

File: var/spack/repos/builtin/packages/py-cython/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-cython/package.py]: Would like to keep the 3.13 upper bound until we know which version will add 3.14 support. When 3.14 is added, we prob won't think to update this package too (unless CI catches an issue).

**Review by haampie — DISMISSED** [review]
looks great, let's see what gitlab thinks (although it never caught that URL issue :thinking:)

**Comment by haampie** [comment]
on the opposite side of things #47719 I noticed py-cffi is also `@:3.11` only, are you looking at that too?

edit: did not see #47713, thanks.

**Comment by haampie** [comment]
Actually, it would be good to merge #47713 with this, or rebase after merging, as it may show more 3.12 related issues in Gitlab.

**Comment by adamjstewart** [comment]
Yes, let's merge either one first (perhaps this one) and then rebase the other.

**Review by haampie — APPROVED** [review]

**Pull request merged**
